### ISSUE TITLE
ReflectionProperty caused Segmentation Fault

### ISSUE THREAD

**Issue by SirSnyder** [krakjoe/pthreads#696]
Hello @krakjoe,
hello @tpunt,

something strange. Concurrent calls of ReflectionProperty->getName() leads to SegmentationFaults.

Faults
```
    class TestClass
    {
        private $property;
    }

    class Segfault extends Thread
    {
        public function run()
        {
            var_dump(__METHOD__, Thread::getCurrentThreadId());

            $reflectionClass = new ReflectionClass(TestClass::class);

            while(true)
            {
                $properties  = $reflectionClass->getProperties();

                foreach($properties as $property)
                {
                    $property->getName();
                }
            }
        }
    }

    $threads = [];

    for($i = 0; $i < 5; ++$i)
    {
        $thread = new Segfault();
        $thread->start();

        $threads[] = $thread;
    }

    foreach($threads as $thread)
    {
        var_dump('join', $thread->getThreadId());

        $thread->join();
    }
```

Sequential access works
```
    class TestClass
    {
        private $property;
    }

    class Segfault extends Thread
    {
        private $share;

        public function __construct(Volatile $share)
        {
            $this->share = $share;
        }

        public function run()
        {
            var_dump(__METHOD__, Thread::getCurrentThreadId());

            $reflectionClass = new ReflectionClass(TestClass::class);

            for($i = 0; $i < 10000; ++$i)
            {
                $properties  = $reflectionClass->getProperties();

                foreach($properties as $property)
                {
                    $property->getName();
                }
            }

            $this->share['done'] = true;

            while(!isset($this->share['quit']));
        }
    }
    $share = new Volatile();

    $threads = [];

    for($i = 0; $i < 5; ++$i)
    {
        unset($share['done']);

        $thread = new Segfault($share);
        $thread->start();

        $threads[] = $thread;

        while(!isset($share['done']));
    }
    $share['quit'] = true;

    foreach($threads as $thread)
    {
        var_dump('join', $thread->getThreadId());

        $thread->join();
    }
```

Synchronized doesn't work
```
class TestClass
    {
        private $property;
    }

    class Segfault extends Thread
    {
        private $mutex;

        public function __construct(Threaded $mutex)
        {
            $this->mutex = $mutex;
        }

        public function run()
        {
            var_dump(__METHOD__, Thread::getCurrentThreadId());

            $reflectionClass = new ReflectionClass(TestClass::class);

            while(true)
            {
                $properties  = $reflectionClass->getProperties();

                foreach($properties as $property)
                {
                    $this->mutex->synchronized(function () use ($property)
                    {
                        $property->getName();
                    });
                }
            }
        }
    }

    $mutex = new Threaded();

    $threads = [];

    for($i = 0; $i < 5; ++$i)
    {
        $thread = new Segfault($mutex);
        $thread->start();

        $threads[] = $thread;
    }

    foreach($threads as $thread)
    {
        var_dump('join', $thread->getThreadId());

        $thread->join();
    }
```
I say "strange" because ReflectionClass and ReflectionProperty are internal classes and should exist separately per thread. Nothing should get shared between them but possibly I'm wrong. What do you guys think? Maybe related to the internal php core string handling?

debian, php 7.1.2 dev and pthreads master

Thx

**Comment by tpunt** [comment]
This is what I believe is happening.

On module startup, the `ReflectionProperty` class [declares two internal properties](https://github.com/php/php-src/blob/01ec14afa4b028a65c4b507babdd4cc6bd0b3395/ext/reflection/php_reflection.c#L6761-L6762) (allocated as [persistent strings](https://github.com/php/php-src/blob/master/Zend/zend_API.c#L3717)). Upon updating the `properties_info` hash table on a ce, an attempt is made to store the property name [as an interned string](https://github.com/php/php-src/blob/01ec14afa4b028a65c4b507babdd4cc6bd0b3395/Zend/zend_API.c#L3705). This fails because `zend_string`s cannot be interned in ZTS mode (well, not until recently in the master branch).

When pthreads prepares classes upon thread setup, it copies property names from the `property_info` struct. This, unfortunately, has no effect in the case of this issue because this is an internal class, and internal classes are not manually copied (they are simply looked up via `zend_lookup_class`). 

This problem seems to specifically revolve around the initialisation of the properties table for an object (basically, whenever an invocation to `rebuild_object_properties` is made). This is problematic because when [it inserts a new property](https://github.com/php/php-src/blob/01ec14afa4b028a65c4b507babdd4cc6bd0b3395/Zend/zend_object_handlers.c#L92) into the properties table, the reference count for that property name [gets incremented](https://github.com/php/php-src/blob/01ec14afa4b028a65c4b507babdd4cc6bd0b3395/Zend/zend_hash.h#L1015). Given that the property name was allocated upon module initialisation, and has not been interned (or otherwise copied by pthreads), it occurs on the originally allocated persistent string.

With multiple threads updating the same piece of memory without any sort of locking, we get data corruption problems (in this case, lost reference incrementing). So this is why the reference count is able to reach 0 when one of the threads performs a `zend_string_release` on the property name. This causes the heap corruption error in ZMM because the string was allocated on the main thread's ZMM heap, and not the currently executing thread's ZMM heap (even if it did successfully release it, other threads attempting to access that property, including the main thread, would segfault).

This lost refcounting update would explain why you cannot observe this problem when either the script is executing sequentially, or when you have explicitly used synchronization.

Now, as for the fix, that's where I'm out of ideas... What I can say, though, is that this problem does not affect the master branch of PHP (soon to be PHP 7.2). This is because (as noted above), `zend_strings` can be interned with ZTS mode enabled, which do's away with the need to reference count strings, thereby avoiding this problem altogether.

**Comment by SirSnyder** [comment]
Thanks for your detailed explanation @tpunt. I've tested the master branch and yes, my error prone examples are working fine with php 7.2-dev. The usage of pthreads in combination with php 7.1 might be critical. May be that should get mentioned in the readme.

**Comment by tpunt** [comment]
@SirSnyder Instead of cautioning pthreads users, finding a solution to this problem would definitely be better. PHP 7.2 is over 6 months away from being released still, so it's unlikely many users will want to (or be able to) use the development version of PHP until then.

I tried last week to make a hard copy of all internal classes per interpreter instance, though I didn't manage to get things working (it just caused an endless amount of problems). Hopefully I'll find some more time next week to have another go at this.

**Comment by SirSnyder** [comment]
@tpunt sure, that would be much better. I tried that too, to copy all internal classes per interpreter instance, but you know the result. Maybe my approach was to simple or naive. I'm open to any idea or approach.

**Comment by SirSnyder** [comment]
@tpunt I've spent the last weekend at this topic. A class copy itself doesn't seem to be a problem. It's more the structure of some internal extensions, as example the Reflection extension. https://github.com/php/php-src/blob/01ec14afa4b028a65c4b507babdd4cc6bd0b3395/ext/reflection/php_reflection.c#L76 macro `METHOD_NOTSTATIC` prevents the possibility to handle different `zend_class_entry`s. Perhaps a solution could be to wrap the internal functions of reflection* classes. A wrapper function could 
- synchronize access 
- swap current `zend_class_entry` pointer https://github.com/php/php-src/blob/01ec14afa4b028a65c4b507babdd4cc6bd0b3395/ext/reflection/php_reflection.c#L65. But I have no idea how to access this pointer out of pthreads.

**Comment by krakjoe** [comment]
This was fixed by 7.2
